### DISCUSSION TITLE
Removing small pixels from scans

### DISCUSSION THREAD

**Discussion by 15frogs** [ImageMagick/ImageMagick#6328] [Help] [answered]
### ImageMagick version

7.1.1-8

### Operating system, version and so on

Win 10

### Description

Hi there,

I am currently working on a project for which I need to scan some technical drawings at very high resolution (1200dpi). Unfortunately, all my scans are covered with lots of small black spots, which I would now like to remove. Should be easy, considering the comparatively small size of the artifacts (usually <80px) and the fact that all scans are black and white. After some research, I found the ImageMagick tool, and have been playing around with the `-connected compontens` feature for the last couple of days, trying to get it to do what I want. It seems though, as if my scans are too large, as I am getting an error message `@ error/vision.c/ConnectedComponentsImage/1091` when trying to remove the artifacts from an entire scan. For a small cutout of the scan (about 1/20 of the original image), the command works pretty much perfectly. However, since manually cutting the images into 20 pieces, then running the command and stitching them back together is just as fast as manually removing the artifacts in paint / gimp / photoshop, I am currently trying to find the reason / fix for the error message.

After doing some more research, I have found discussion #5163 , where someone mentioned that the Q16 version limits the command to 65538 regions, however my example image does not even have 65k pixels, let alone 65k regions. I tried to compile a Q32 version to gain more regions, but I'm getting compile errors in Visual Studio and at this point, I have already spent way too much time trying to fix this problem to justify going on with the magick approach. My last hope is that someone here might've already encountered this issue and knows a solution or can give me any info on why this is occurring. I am currently working with the following command and example picture: 
 
`magick in.png -define connected-components:area-threshold=80 -connected-components 4 -auto-level out.png` (Taken from here: https://imagemagick.org/script/connected-components.php) 

![in](https://github.com/ImageMagick/ImageMagick/assets/133606306/61795b78-0d3d-4a55-aea7-42d734089e23)

I tried both version 7.1.1-8-Q16-HDRI and 7.1.1-8-Q16-HDRI (Win 10 Home) without success.

Any help would help me out immensely and is much appreciated!
Thanks

**Comment by fmw42** [comment]
More gray levels in the image the more regions you will have.  Make your image binary. Then filter that. Then use the result as a mask to remove the artifacts in the original version.

**Reply by fmw42** [reply]
Or use Q32 IM compile to get more regions.

**Reply by urban-warrior** [reply]
The maximum # of objects in connected components is 65536 even for Q32.

**Comment by urban-warrior** [comment]
Could you not use noise reduction to remove most of the salt-n-pepper artifacts then use connected components?

**Comment by fmw42** [comment, marked as answer]
Unix syntax:

```
magick schematic.png \
\( -clone 0 -fill white -colorize 100 \) \
\( -clone 0 -threshold 50% \
-define connected-components:area-threshold=100 \
-define connected-components:mean-color=true \
-connected-components 8 \) \
-compose over -composite \
schematic_result.png
```

Windows syntax:

```
magick schematic.png ^
( -clone 0 -fill white -colorize 100 ) ^
( -clone 0 -threshold 50% ^
-define connected-components:area-threshold=100 ^
-define connected-components:mean-color=true ^
-connected-components 8 ) ^
-compose over -composite ^
schematic_result.png

```

If in a .bat file, double the % to %%

Increase/decrease the area threshold (area in pixels) for the size of the areas to be removed to remove more/less regions.  Mean color = true tells it you want gray level output (average color of regions) rather than ID values of regions.

Result:

![schematic_result](https://github.com/ImageMagick/ImageMagick/assets/13458218/fc4b82f3-1ede-4a72-8748-3fc74c8c65bb)

**Reply by 15frogs** [reply]
You are a saint. Thank you for the detailed explanation and directly giving me the code as well, this way I don't need to put anymore work in, while knowing how it works :) You just saved me countless hours of work, manually formatting the images. Thank you very much!
